File: unity/Assets/Scripts/unityagent/SystemMessage/EgressSystemMessageBuffer.cs
using System.Collections.Generic;

namespace UnityAgent.SystemMessage
{
public class EgressSystemMessageBuffer
{
    private uint bufferSize = 10;
    private readonly bool dropNew = false;
    private readonly LinkedList<EgressSystemMessage> buffer = new();

    public EgressSystemMessageBuffer()
    {
    }

    public EgressSystemMessageBuffer(uint bufferSize, bool dropNew)
    {
        BufferSize = bufferSize;
        this.dropNew = dropNew;
    }

    public uint BufferSize
    {
        get => bufferSize;
        set
        {
            if (value == 0)
                bufferSize = 10;
        }
    }
    public bool DropNew => dropNew;


    public void Enqueue(EgressSystemMessage message)
    {
        lock (buffer)
        {
            if (HasFreeSpace())
                buffer.AddLast(message);
            else if (DropsOld())
            {
                buffer.RemoveFirst();
                buffer.AddLast(message);
            }
        }
    }

    public EgressSystemMessage Dequeue()
    {
        lock (buffer)
        {
            EgressSystemMessage esm = buffer.First.Value;
            buffer.RemoveFirst();
            return esm;
        }
    }

    public void Clear()
    {
        buffer.Clear();
    }

    private bool DropsOld()
    {
        return !dropNew;
    }

    private bool HasFreeSpace()
    {
        return buffer.Count < bufferSize;
    }
}
}

File: ros/ros_packages/rosagent/rosagent/message/egress_system_message_buffer.py
from collections import deque

class EgressSystemMessageBuffer:

    def __init__(self, size=10, drop_new=True):
        self._buffer = deque(maxlen=size)
        self._size = size
        self._drop_new = drop_new
    
    def enqueue(self, message):
        if self.can_enqueue_item():
            self._buffer.append(message)
            return True
        return False
    
    def dequeue(self):
        try:
            return self._buffer.popleft()
        except IndexError:
            return None

    def get_num_items(self):
        return len(self._buffer)

    def get_size(self):
        return self._size

    def is_full(self):
        return not self._has_free_space()

    def can_enqueue_item(self):
        return self._drop_old() or self._has_free_space()

    def _drop_old(self):
        return not self._drop_new

    def _has_free_space(self):
        return len(self._buffer) < self._size
    